### PR TITLE
Refactor PSO

### PR DESCRIPTION
**Pull request by jinyongyoo**

### CODE REVIEW

File: textattack/search_methods/particle_swarm_optimization.py [hunk 1]
```diff
@@ -24,252 +25,313 @@ class ParticleSwarmOptimization(SearchMethod):
...
+                        population[k],
+                        turn_prob,
+                        initial_result.attacked_text,
+                    )
+
+            # Check if there is any successful attack in the current population
+            pop_results, self._search_over = self.get_goal_results(
+                [p.attacked_text for p in population]
+            )
+            if self._search_over:
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/particle_swarm_optimization.py:275]: Yeah that's true.

File: textattack/search_methods/population_based_method.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+from abc import ABC, abstractmethod
+
+from textattack.search_methods import SearchMethod
+
+
+class PopulationBasedMethod(SearchMethod, ABC):
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/population_based_method.py:6]: Can you explain more what you mean?

File: textattack/search_methods/particle_swarm_optimization.py [hunk 3]
```diff
@@ -24,252 +25,313 @@ class ParticleSwarmOptimization(SearchMethod):
...
+
+    def check_transformation_compatibility(self, transformation):
+        """The genetic algorithm is specifically designed for word
+        substitutions."""
+        return transformation_consists_of_word_swaps(transformation)
+
+    def extra_repr_keys(self):
+        return ["pop_size", "max_iters", "post_turn_check", "max_turn_retries"]
+
+
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/particle_swarm_optimization.py:322]: Yes we can move `sigmoid` to `utils` and `count_change_ratio` to `AttackedText`, but I would just keep `normalize` where it is since normalization can depend on the specific method.

File: textattack/search_methods/improved_genetic_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+        temp=0.3,
+        max_replace_times_per_index=5,
+        give_up_if_no_improvement=False,
+        post_crossover_check=True,
+        max_crossover_retries=20,
+    ):
+        self.max_iters = max_iters
+        self.pop_size = pop_size
+        self.temp = temp
+        self.max_replace_times_per_index = max_replace_times_per_index
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/improved_genetic_algorithm.py:46]: That's because of a typo. Good catch

File: textattack/search_methods/improved_genetic_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            if population[0].result.score > current_score:
+                current_score = population[0].result.score
+            elif self.give_up_if_no_improvement:
+                break
+
+            pop_scores = torch.Tensor([pm.result.score for pm in population])
+            logits = ((-pop_scores) / self.temp).exp()
+            select_probs = (logits / logits.sum()).cpu().numpy()
+
+            parent1_idx = np.random.choice(pop_size, size=pop_size - 1, p=select_probs)
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/improved_genetic_algorithm.py:236]: That's something the author for IGA wrote himself. IGA initializes population by replacement each word with its best replacement. So we our original text is 20 words long, then we have population of 20, despite setting `self.pop_size=60`. If original text is 80 words long, then we cut it off to 60 according to `self.pop_size`

File: textattack/shared/attack.py [hunk 6]
```diff
@@ -86,15 +105,31 @@ def __init__(
...
         of possible transformations through the applicable constraints.
 
         Args:
             current_text: The current ``AttackedText`` on which to perform the transformations.
             original_text: The original ``AttackedText`` from which the attack started.
-            apply_constraints: Whether or not to apply post-transformation constraints.
+        Returns:
+            A filtered list of transformations where each transformation matches the constraints
+        """
+        transformed_texts = self.transformation(
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/shared/attack.py:118]: I removed it because I don't think there's really anything to benefit from using np.array (but correct me if I'm wrong). Numpy arrays are fast for matrix and math-y operations but we have a list of AttackedText, which is an object with varying size. I really don't think creating an numpy array and copying the data again to fit into a contiguous memory block is that beneficial, especially when we'll be filtering stuff again using constraints.

File: textattack/transformations/transformation.py [hunk 7]
```diff
@@ -60,3 +60,12 @@ def extra_repr_keys(self):
         return []
 
     __repr__ = __str__ = default_class_repr
+
+
+class RandomTransformation(ABC):
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/transformations/transformation.py:65]: Well the main benefit is that we don't have to edit every transformation class.

File: textattack/search_methods/particle_swarm_optimization.py [hunk 8]
```diff
@@ -24,252 +25,313 @@ class ParticleSwarmOptimization(SearchMethod):
...
         self.max_iters = max_iters
         self.pop_size = pop_size
-        self.search_over = False
+        self.post_turn_check = post_turn_check
+        self.max_turn_retries = 20
 
-        self.Omega_1 = 0.8
-        self.Omega_2 = 0.2
+        self._search_over = False
+        self.omega_1 = 0.8
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/particle_swarm_optimization.py:45]: I think PEP doesn't like capitalization of variables?

File: textattack/search_methods/particle_swarm_optimization.py [hunk 9]
```diff
@@ -24,252 +25,313 @@ class ParticleSwarmOptimization(SearchMethod):
...
+            original_text (AttackedText): Original text for constraint check if `self.post_turn_check=True`.
+        Returns:
+            New `Position` that we moved to (or if we fail to move, same as `source_text`)
+        """
+        assert len(source_text.words) == len(
+            target_text.words
+        ), "Word length mismatch for turn operation."
+        assert len(source_text.words) == len(
+            prob
+        ), "Length mistmatch for words and probability list."
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/particle_swarm_optimization.py:93]: Good catch

File: textattack/search_methods/particle_swarm_optimization.py [hunk 10]
```diff
@@ -24,252 +25,313 @@ class ParticleSwarmOptimization(SearchMethod):
...
-            prob_list.append(p)
+        if random_result == pop_member.result:
+            return False
+        else:
+            pop_member.attacked_text = random_result.attacked_text
+            pop_member.result = random_result
+            return True
 
-        prob_list = self._norm(prob_list)
+    def _equal(self, a, b):
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/particle_swarm_optimization.py:74]: That is from the original code base. Is there an alternative name you think is better?

File: textattack/transformations/transformation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/transformations/transformation.py:65]: On second thought, I think property would work (and is better).

File: textattack/search_methods/particle_swarm_optimization.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/search_methods/particle_swarm_optimization.py:45]: maybe change the capitalization of the other variables too then?

File: textattack/search_methods/particle_swarm_optimization.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/search_methods/particle_swarm_optimization.py:74]: I don't really understand what it's doing, but I feel like I would think an `_equal` function which takes 2 inputs would return a boolean determining whether they are equal, rather than a number based on if they are equal.

File: textattack/search_methods/population_based_method.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/search_methods/population_based_method.py:6]: Just changing the `PopulationBasedMethod` to `PopulationBasedSearch`. We named greedy/beam `GreedySearch` and `BeamSearch`, rather than `GreedyMethod` and `BeamMethod`.

File: textattack/search_methods/genetic_algorithm.py [hunk 11]
```diff
@@ -133,144 +120,107 @@ def _crossover(self, pop_member1, pop_member2, original_result):
...
+                iter(transformed_text.attack_attrs["newly_modified_indices"])
             )
+            num_replacements_per_word[diff_idx] += 1
+
+        # Just b/c there are no replacements now doesn't mean we never want to select the word for perturbation
+        # Therefore, we give small non-zero probability for words with no replacements
+        # Epsilon is some small number to approximately assign 1% probability
+        num_total_candidates = np.sum(num_replacements_per_word)
+        epsilon = max(1, int(num_total_candidates * 0.01))
+        for i in range(len(num_replacements_per_word)):
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/genetic_algorithm.py:200]: But that if `num_replacements_per_word[i] < epsilon`, then we are effectively overwriting it with epsilon. The point of epsilon is to deal with cases where `num_replacements_per_word[i] == 0`. I think we should set epsilon to be 10% of minimum of `num_replacements_per_word` (or 1). Then, we can safely do `num_replacements_per_word = np.max(num_replacements_per_word,epsilon)`

File: textattack/search_methods/genetic_algorithm.py [hunk 12]
```diff
@@ -307,6 +261,8 @@ def _perform_search(self, initial_result):
                 # `crossover` method and `perturb` method.
                 if self._search_over:
                     break
+            if self._search_over:
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/genetic_algorithm.py:264]: good point

File: textattack/shared/attack.py [hunk 13]
```diff
@@ -74,6 +74,25 @@ def __init__(
             else:
                 self.constraints.append(constraint)
 
+        # Check if we can use transformation cache for our transformation.
+        if isinstance(self.transformation, RandomTransformation) or (
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/shared/attack.py:78]: Not sure. Validators are kinda supposed to do hard checks (i.e. raise ValueError), but in this case we just want to check if we can use caching.

File: textattack/search_methods/population_based_method.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/population_based_method.py:6]: ahh I see I misread your comment

File: textattack/search_methods/particle_swarm_optimization.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/particle_swarm_optimization.py:74]: Screen Shot 2020-07-25 at 6 19 44 PM

This is the actual function it's supposed to be... I'm not sure why return max velocity, however. 
I agree that `_equal` isn't fitting, but, I don't think it's a huge issue. If you still want to rename it to something else, I'll be more than happy to change it.

File: textattack/search_methods/genetic_algorithm.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/search_methods/genetic_algorithm.py]: yeah this solution makes sense

File: tests/test_command_line/test_attack.py [hunk 14]
```diff
@@ -119,7 +119,7 @@
         "run_attack_faster_alzantot_recipe",
         (
             "textattack attack --model lstm-mr --recipe faster-alzantot --num-examples 3 "
-            "--num-examples-offset 20 --shuffle=False"
+            "--num-examples-offset 32 --shuffle=False"
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply tests/test_command_line/test_attack.py:122]: For offset 20, 2 example are skipped, which is useless for testing.

File: textattack/search_methods/genetic_algorithm.py [hunk 15]
```diff
@@ -99,178 +96,123 @@ def _perturb(self, pop_member, original_result, specified_idx=-1):
...
                 ]
-                filtered = self.filter_transformations(
-                    [new_text], x2_text, original_text=original_result.attacked_text
+
+            if not self.post_crossover_check or (
+                new_text.text == x1_text.text or new_text.text == x2_text.text
+            ):
+                break
+
+            if "last_transformation" in new_text.attack_attrs:
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/genetic_algorithm.py:165]: Good idea

File: textattack/search_methods/improved_genetic_algorithm.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        Initialize a population of size `pop_size` with `initial_result`
+        Args:
+            initial_result (GoalFunctionResult): Original text
+            pop_size (int): size of population
+        Returns:
+            population as `list[PopulationMember]`
+        """
+        words = initial_result.attacked_text.words
+        # For IGA, `num_replacements_per_word` represents the number of times the word at each index can be modified
```
[the remark this answers is not in the document]
Comment (jinyongyoo) [reply textattack/search_methods/improved_genetic_algorithm.py:99]: Yeah we used the same variable to minimize code difference in original GA + IGA, but you're right. it's better to have different names.

File: textattack/search_methods/genetic_algorithm.py [hunk 17]
```diff
@@ -104,23 +103,52 @@ def _perturb(self, pop_member, original_result, index=None):
...
+
+        Args:
+            new_text (AttackedText): Text produced by crossover operation
+            parent_text1 (AttackedText): Parent text of `new_text`
+            parent_text2 (AttackedText): Second parent text of `new_text`
+            original_text (AttackedText): Original text
+        Returns:
+            `True` if `new_text` meets the constraints. If otherwise, return `False`.
+        """
+        if "last_transformation" in new_text.attack_attrs:
```
Comment (uvafan) [inline textattack/search_methods/genetic_algorithm.py:138]: This constraint checking logic still seems confusing to me. Why do we not check constraints if the 2 parent texts haven't been previously transformed?

File: textattack/search_methods/genetic_algorithm.py [hunk 18]
```diff
@@ -61,19 +61,18 @@ def _perturb(self, pop_member, original_result, index=None):
...
             return pop_member
         iterations = 0
         while iterations < non_zero_indices:
             if index:
                 idx = index
             else:
-                w_select_probs = num_replacements_per_word / np.sum(
-                    num_replacements_per_word
-                )
+                w_select_probs = word_attr_list / np.sum(word_attr_list)
```
Comment (uvafan) [inline textattack/search_methods/genetic_algorithm.py:75]: Instead of doing the `word_attr_list` things depending on `self.atr_name`, I think we should:
- Get rid of the `self.attr_name` thing, it feels like not great practice and hard to read/understand.
- Make getting the selection probs into a separate function so it can be overridden by the `Alzantot` and `Improved` versions.
- Make all `PopulationMember`s have both of the attribute lists, but which attribute gets used depends on the specifics of select probs as described above.

File: textattack/search_methods/genetic_algorithm.py (hunk 17 above)
Comment (jinyongyoo) [reply textattack/search_methods/genetic_algorithm.py:138]: If both parents texts haven't been transformed, then both are equivalent to the original text. Therefore, the child must also be equivalent to the original text, meaning the constraint check is not necessary. We could do the check, but checking constraint requires `last_transformation` to exist in the `attack_attrs` of the transformed text.

File: textattack/search_methods/genetic_algorithm.py (hunk 18 above)
Comment (jinyongyoo) [reply textattack/search_methods/genetic_algorithm.py:75]: I really like those changes, but one problem is that `word_attr_list` is actually a copy of the attribute list and used as temporary variable within the while loop.  So it's unclear how this works with the word selection prob function. We can't create copy of the attribute list within the function and use it since its scope is not just limited to the function.
